### PR TITLE
Auto ECR Companion Stack Deletion

### PR DESCRIPTION
**Pull request by hnnasit**
#### Why is this change necessary?
Allows users to delete auto ecr companion stacks
#### Checklist

- [x] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [x] Write unit tests
- [ ] Write/update functional tests
- [ ] Write/update integration tests
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

File: samcli/commands/delete/delete_context.py [hunk 1]
```diff
@@ -137,6 +155,85 @@ def guided_prompts(self):
...
+    def ecr_companion_stack_prompts(self):
+        """
+        User prompt to delete the ECR companion stack.
+        """
+        click.echo(f"\tFound ECR Companion Stack {self.companion_stack_name}")
+        if not self.no_prompts:
+            delete_ecr_companion_stack_prompt = confirm(
+                click.style(
+                    "\tDo you you want to delete the ECR companion stack"
+                    + f" {self.companion_stack_name} in the region {self.region} ?",
```
Comment (hawflau) [inline samcli/commands/delete/delete_context.py:167]: bit: `+` is not needed here. Two strings can just concatenate without `+`

File: samcli/lib/delete/cf_utils.py [hunk 2]
```diff
@@ -16,6 +16,7 @@
 class CfUtils:
     def __init__(self, cloudformation_client):
         self._client = cloudformation_client
+        # self._resource_client = cloudformation_resource_client
```
Comment (hawflau) [inline samcli/lib/delete/cf_utils.py:19]: Please remove the line if it is not needed

File: samcli/lib/package/ecr_uploader.py [hunk 3]
```diff
@@ -114,35 +113,48 @@ def delete_artifact(self, image_uri: str, resource_id: str, property_name: str):
                 # Image not found
                 image_details = resp["failures"][0]
                 if image_details["failureCode"] == "ImageNotFound":
-                    LOG.error("ImageNotFound Exception")
-                    message_fmt = (
-                        "Could not delete image for {property_name}"
-                        " parameter of {resource_id} resource as it does not exist. \n"
+                    LOG.debug(
+                        "Could not delete image for %s" " parameter of %s resource as it does not exist. \n",
```
Comment (hawflau) [inline samcli/lib/package/ecr_uploader.py:117]: nit: the two strings can be merged into one

File: samcli/lib/package/packageable_resources.py [hunk 4]
```diff
@@ -517,6 +518,27 @@ class CloudFormationResourceVersionSchemaHandlerPackage(ResourceZip):
     PROPERTY_NAME = RESOURCES_WITH_LOCAL_PATHS[AWS_CLOUDFORMATION_RESOURCEVERSION][0]
 
 
+class ECRResource(Resource):
```
Comment (hawflau) [inline samcli/lib/package/packageable_resources.py:521]: Can you add a docstring explaining what this class is for?
Seems like this class is not for exporting at all. Only for deleting

File: samcli/lib/package/ecr_uploader.py [hunk 5]
```diff
@@ -114,35 +113,48 @@ def delete_artifact(self, image_uri: str, resource_id: str, property_name: str):
...
+        :param: physical_id of the repository to be deleted
+        """
+        try:
+            click.echo(f"\t- Deleting ECR repository {physical_id}")
+            self.ecr_client.delete_repository(repositoryName=physical_id, force=True)
+        except self.ecr_client.exceptions.RepositoryNotFoundException:
+            # If the repository is empty, cloudformation automatically deletes
+            # the repository when cf_client.delete_stack is called.
+            LOG.debug("Could not find repository %s", physical_id)
+            pass
```
Comment (hawflau) [inline samcli/lib/package/ecr_uploader.py:152]: This seems to be unnecessary

File: samcli/commands/delete/delete_context.py [hunk 6]
```diff
@@ -27,6 +35,9 @@
 
 
 class DeleteContext:
+
+    ecr_repos: Dict[str, Dict[str, str]]
```
Comment (moelasmar) [inline samcli/commands/delete/delete_context.py:39]: why this variable is a class level variable?

File: samcli/commands/delete/delete_context.py [hunk 7]
```diff
@@ -69,22 +82,28 @@ def parse_config_file(self):
...
                 self.stack_name = config_options.get("stack_name", None)
             # If the stack_name is same as the one present in samconfig file,
-            # get the information about parameters if not specified by customer.
+            # get the information about parameters if not specified by user.
             if self.stack_name and self.stack_name == config_options.get("stack_name", None):
                 LOG.debug("Local config present and using the defined options")
                 if not self.region:
                     self.region = config_options.get("region", None)
-                    click.get_current_context().region = self.region
+                    Context.get_current_context().region = self.region
```
Comment (moelasmar) [inline samcli/commands/delete/delete_context.py:90]: do we need to double check that the region we get it from the config file is not none before updating the context ?

File: samcli/commands/delete/delete_context.py [hunk 8]
```diff
@@ -69,22 +82,28 @@ def parse_config_file(self):
...
                 if not self.profile:
                     self.profile = config_options.get("profile", None)
-                    click.get_current_context().profile = self.profile
+                    Context.get_current_context().profile = self.profile
```
Comment (moelasmar) [inline samcli/commands/delete/delete_context.py:93]: same as the previous comment

File: samcli/commands/delete/delete_context.py (hunk 7 above)
Comment (hnnasit) [reply samcli/commands/delete/delete_context.py:90]: That makes sense, context should be updated after profile and region are assigned values other than None.

**Review by moelasmar — APPROVED** [review]

**Review by hawflau — APPROVED** [review]

**Pull request merged**